File: leetcode/0239-sliding-window-maximum/solution.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        Deque<Integer> dq = new ArrayDeque<>();
        int n= nums.length;
        int[] result = new int[n-k+1];

        for(int curr_ptr=0;curr_ptr<n;curr_ptr++)
        {
            while(!(dq.isEmpty()) && dq.peekFirst()<curr_ptr-k+1)
            {
                dq.pollFirst();
            }
            while(!(dq.isEmpty()) && nums[dq.peekLast()]<nums[curr_ptr])
            {
                dq.pollLast();
            }
            dq.offerLast(curr_ptr);
            if(curr_ptr>=k-1)
            {
                result[curr_ptr-k+1] = nums[dq.peekFirst()];
            }
        }

        return result;
    }
}

File: leetcode/0239-sliding-window-maximum/solution.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> ans;
        multiset<int> ms;
        int l = 0, r = 0;
        while(r<nums.size()){
            ms.insert(nums[r]);
            
            if(r - l + 1 == k){
                ans.push_back(*ms.rbegin());
                ms.erase(ms.find(nums[l]));
                l++;
            }
            r++;
        }
        return ans;
    }
};
